src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+    /// Errors returned by this method are intentionally sparse, for simplicity of implementation.
+    /// It is expected that only buggy frontends or fuzzers should ever emit IR that fails this
+    /// validation pass.
+    pub(super) fn validate_module_handles(
+        module: &crate::Module,
+    ) -> Result<(), InvalidHandleError> {
+        let &crate::Module {
+            ref constants,
+            ref entry_points,
+            ref functions,
+            ref global_variables,
+            ref types,
+        } = module;
+
+        // NOTE: Types being first is important. All other forms of validation depend on this.
+        for (this_handle, type_) in types.iter() {
+            let &crate::Type { name: _, ref inner } = type_;
+
+            match inner {
+                &crate::TypeInner::Scalar { .. }
+                | &crate::TypeInner::Vector { .. }
+                | &crate::TypeInner::Matrix { .. }
+                | &crate::TypeInner::ValuePointer { .. }
+                | &crate::TypeInner::Atomic { .. }
+                | &crate::TypeInner::Image { .. }
+                | &crate::TypeInner::Sampler { .. } => (),
+                &crate::TypeInner::Pointer { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Array { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Struct { ref members, .. } => {
+                    this_handle.check_dep_iter(members.iter().map(|m| m.ty))?;
+                }
+                &crate::TypeInner::BindingArray { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+            }
+        }
+
+        let validate_type = |handle: Handle<crate::Type>| handle.check_valid_for_uniq(types);
+
+        for (this_handle, constant) in constants.iter() {
+            let &crate::Constant {
+                name: _,
+                specialization: _,
+                ref inner,
+            } = constant;
+            match *inner {
+                crate::ConstantInner::Scalar { .. } => (),
+                crate::ConstantInner::Composite { ty, ref components } => {
+                    validate_type(ty)?;
+                    this_handle.check_dep_iter(components.iter().copied())?;
+                }
+            }
+        }
+
+        let validate_constant = |handle: Handle<crate::Constant>| handle.check_valid_for(constants);
+
+        for (_handle, global_variable) in global_variables.iter() {
+            let &crate::GlobalVariable {
+                name: _,
+                space: _,
+                binding: _,
+                ty,
+                init,
+            } = global_variable;
+            validate_type(ty)?;
+            if let Some(init_expr) = init {
+                validate_constant(init_expr)?;
+            }
+        }
+
+        let validate_expressions = |expressions: &Arena<crate::Expression>,
```

Comment:
Could we just inline this into `validate_function`?

Reply:
~~Sure! Changed with 3412651ac5ccfbf5a1a5e5158cf0c00760fe1ff9.~~ EDIT: Reverted with 202d699a06c6607c57980d727a0bc299b8a77caf.

Actually, the inner `match` here is still something we'll want to re-use. We can inline the looping logic (which I did with 9ce5abba). After moving the closure to an associated function (see 23637853), I did end up re-using it with 75654ac7 (see also the convo we have about statements not getting checked [here](https://github.com/gfx-rs/naga/pull/2090#discussion_r1047296957)).